### PR TITLE
QUICKSTEP-36 Quickstep gives segmentation fault when it imports table from a file that does not have access rights.

### PR DESCRIPTION
**Pull request by tarunbansal**
This PR fixes QUICKSTEP-36 
 
- Perform access check on a file before opening it to import a table.

### CODE REVIEW

File: relational_operators/TextScanOperator.cpp [hunk 1]
```diff
@@ -91,6 +91,12 @@ bool TextScanOperator::getAllWorkOrders(
   if (blocking_dependencies_met_ && !work_generated_) {
     for (const std::string &file : files) {
       // Use standard C libary to retrieve the file size.
+
+      // Check file permissions before trying to open it
+      int rval = access(file.c_str(), R_OK);
```
Comment (hbdeshmukh) [inline relational_operators/TextScanOperator.cpp]: Hi @tarunbansal - This is a neat change. 

The ``access`` function may not be available on all the platforms that are used in Quickstep. On unix, it comes from the header ``unistd.h``. 

To solve such issues, typically we check the availability of this header file using CMake. If the function is available, we create a variable, e.g. in this case it could be QUICKSTEP_HAVE_UNISTD and guard this line with ``#ifdef QUICKSTEP_HAVE_UNISTD`` and ``#endif QUICKSTEP_HAVE_UNISTD``. You can dig around in the code to find such usage patterns.

**Comment by tarunbansal** [comment]
Thanks @hbdeshmukh  for noticing this possible ambiguity. I am still digging around the code and will update the PR once the fix is ready.

File: relational_operators/RelationalConfig.h.in [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ **/
+
+#cmakedefine QUICKSTEP_HAVE_UNISTD
```
Comment (zuyu) [inline relational_operators/RelationalConfig.h.in]: Code style: miss the end-of-line symbol.

File: relational_operators/TextScanOperator.cpp [hunk 3]
```diff
@@ -91,6 +91,15 @@ bool TextScanOperator::getAllWorkOrders(
   if (blocking_dependencies_met_ && !work_generated_) {
     for (const std::string &file : files) {
       // Use standard C libary to retrieve the file size.
+
+      // Check file permissions before trying to open it
+      #ifdef QUICKSTEP_HAVE_UNISTD
```
Comment (zuyu) [inline relational_operators/TextScanOperator.cpp]: Hi @tarunbansal Welcome to the team, and please get familiar with our code style at https://cwiki.apache.org/confluence/display/QUICKSTEP/Quickstep+Coding+Standards.

Also, I noticed that you made the file mode from `644` to `755`. Please revert this change first.

Finally, I would have written the code as following:
```
// Add in Line 21.
#include "relational_operators/RelationalConfig.h"  // For QUICKSTEP_HAVE_UNISTD.

#ifdef QUICKSTEP_HAVE_UNISTD
#include <unistd.h>
#endif  // QUICKSTEP_HAVE_UNISTD

// Start from Line 95.
#ifdef QUICKSTEP_HAVE_UNISTD
      // Check file permissions before trying to open it.
      const int access_result = access(file.c_str(), R_OK);
      CHECK_EQ(0, access_result)
          << "File " <<  file << " is not readable due to permission issues.";
#endif  // QUICKSTEP_HAVE_UNISTD
```

**Comment by tarunbansal** [comment]
Hi @zuyu Thanks for the welcome and the information about coding style. I have made the suggested changes and reverted file permission change which was done erroneously in the last commit by me.

File: relational_operators/RelationalConfig.h.in (hunk 2 above)
Comment (hbdeshmukh) [inline relational_operators/RelationalConfig.h.in]: To be consistent with the similar files in other modules, can we rename this file to ``RelationalOperatorsConfig.h.in``?

File: relational_operators/TextScanOperator.cpp [hunk 4]
```diff
@@ -18,6 +18,11 @@
  **/
 
 #include "relational_operators/TextScanOperator.hpp"
+#include "relational_operators/RelationalConfig.h"
```
Comment (hbdeshmukh) [inline relational_operators/TextScanOperator.cpp]: Can you sort the above two included files alphabetically? We follow a convention that lists of included files are sorted. If you are dealing with a large list of included files, you may use your IDE's sorting feature, instead of manual effort.

**Comment by hbdeshmukh** [comment]
Hi @tarunbansal 

I have given some comments. After you address them, I will merge this PR. Thanks.

File: relational_operators/TextScanOperator.cpp (hunk 4 above)
Comment (zuyu) [inline relational_operators/TextScanOperator.cpp]: @hbdeshmukh Alternatively, I suggest to add an empty line between Line 20 and 21.

Comment (zuyu) [inline relational_operators/TextScanOperator.cpp]: @tarunbansal As I suggested before, please add comments `// For QUICKSTEP_HAVE_UNISTD.` after the head file.

File: relational_operators/TextScanOperator.cpp [hunk 5]
```diff
@@ -91,6 +96,14 @@ bool TextScanOperator::getAllWorkOrders(
   if (blocking_dependencies_met_ && !work_generated_) {
     for (const std::string &file : files) {
       // Use standard C libary to retrieve the file size.
+
+#ifdef QUICKSTEP_HAVE_UNISTD
+      // Check file permissions before trying to open it.
+      const int access_result = access(file.c_str(), R_OK);
+      CHECK_EQ(0, access_result)
+          << "File " <<  file << " is not readable due to permission issues.";
```
Comment (zuyu) [inline relational_operators/TextScanOperator.cpp]: Please remove one of two whitespaces before `file`.

File: relational_operators/TextScanOperator.cpp [hunk 6]
```diff
@@ -19,6 +19,12 @@
 
 #include "relational_operators/TextScanOperator.hpp"
 
+#include "relational_operators/RelationalOperatorsConfig.h"  // For QUICKSTEP_HAVE_UNISTD.
```
Comment (hbdeshmukh) [inline relational_operators/TextScanOperator.cpp]: Can you move this include between line 39 and 50, i.e. in the block of included files from quickstep modules. When you do that, make sure the list is sorted.

Comment (zuyu) [inline relational_operators/TextScanOperator.cpp]: @hbdeshmukh No, this line is ok, as the next line immediately uses `QUICKSTEP_HAVE_UNISTD` defined in the header file.

See https://github.com/apache/incubator-quickstep/blob/master/storage/StorageManager.cpp#L20 for more info and examples.

Comment (hbdeshmukh) [inline relational_operators/TextScanOperator.cpp]: Ah, good catch. Thanks @zuyu

**Comment by zuyu** [comment]
@tarunbansal Before we merge this PR, please squash all the commits into one. Thanks!

https://cwiki.apache.org/confluence/display/QUICKSTEP/Workflow+For+Contributors

**Comment by tarunbansal** [comment]
Thanks @hbdeshmukh and @zuyu for the suggestions and help. I have squashed all the commits into one. Please merge this PR if it is good to go now.

**Comment by pateljm** [comment]
Thanks @zuyu and @hbdeshmukh for excellent mentorship on this PR!

**Comment by zuyu** [comment]
LGTM. Merged.
